import 'std.errors'

type File(raw_handle) {
	fun write(data) {
		result := fs.__internal_write(self.raw_handle, data)

		if result == 0 {
			return result.ok(null)
		}

		return result.err(result)
	}
	
    fun read_to_string() {
        data := fs.__internal_read_to_string(self.raw_handle)

        if data == null {
            return result.err(1)
        }

        return result.ok(data)
    }
}

unit fs {
    native __internal_open -> 'fs@open'
    native __internal_create -> 'fs@create'
    native __internal_write -> 'fs@write'
    native __internal_close -> 'fs@close'
    native __internal_read_to_string -> 'fs@read_to_string'

	fun create(filename) {
		file_hdl := __internal_create(filename)

		if file_hdl == null {
			return result.err(1)
		}

		return result.ok(new File(file_hdl))
	}

    fun open(filename) {
        file_hdl := __internal_open(filename)

        if file_hdl == null {
            return result.err(1)
        }

        return result.ok(new File(file_hdl))
    }
}
